added support for error_info r-value references

rnburn opened the pull request:
adds support to transport exception data using move semantics

Example:

```
boost::exception e;
typedef boost::error_info<struct tag,Data> ErrorInfo;
Data data;
e << ErrorInfo(std::move(data));
throw e;
```

apolukhin commented on include/boost/exception/info.hpp:
```diff
@@ -175,6 +185,22 @@ boost
+        set_info( E const & x, error_info<Tag,T> && v )
+            {
+            typedef error_info<Tag,T> error_info_tag_t;
+            shared_ptr<error_info_tag_t> p( new error_info_tag_t(std::move(v)) );
```
use `shared_ptr<error_info_tag_t> p = make_shared<error_info_tag_t>(std::move(v))`
It calls less memory allocations

apolukhin commented on include/boost/exception/detail/error_info_impl.hpp:
```diff
@@ -46,6 +47,13 @@ boost
         error_info( value_type const & value );
+#ifndef BOOST_NO_CXX11_RVALUE_REFERENCES
+        error_info( value_type && value);
+        error_info( error_info const & ) = default;
```
Check for `BOOST_NO_CXX11_DEFAULTED_FUNCTIONS` and define function bodies by hand in worst case.

MarcelRaad replied:
Also, Visual C++ 12 doesn't support defaulted move constructors and move assignment operators even though it supports defaulted functions in general.

apolukhin commented on include/boost/exception/info.hpp:
```diff
@@ -46,6 +46,16 @@ boost
+    inline
+    error_info<Tag,T>::
+    error_info( value_type && value ):
+        value_(std::move(value))
```
Add `#include <utility>` for `std::move`

rnburn commented:
I removed the default constructor and assignment operators (just implemented all of them to avoid the problem with Visual C++ 12); changed to use make_shared; and added the missing header files.

apolukhin commented:
Looks like this was fixed in 5a040eedf54691d202e3099033d6d1c5c88a736e

jeking3 commented:
@zajo perhaps is this was already fixed, the PR could be closed?
